### ISSUE TITLE
Allow messages to be registered from any SQL function

### ISSUE THREAD

**Issue by demianuco** [graphile/operation-hooks#12]
One of the great things about this plugin is that it allows pre/post mutation functions to "register" one or more notification messages that can be passed back to the GraphQL client. Crucially, these messages **augment** (not replace) the mutation response, which means that the mutation can still return what the client expects (e.g. the new entity in the case of an insert) PLUS any optional relevant notification messages.

It would be great if this convenience could be extended to ALL mutation functions. In particular, I'd like for any database function to be able to register one or more notification messages to an "ambient" transaction-specific context that could ultimately be read by PostGraphile (at the end of the request/response lifecycle, when generating the GraphQL response) to add those messages to the response, just like it currently does with the operation hooks functions.

Messages could be added via a universally accessible convenience function accepting the same four properties of the current notification type. For example, `add_notification(type, message, code, path)` would add a notification message to the "ambient" transaction-specific context.

Ideally, messages could also be checked via similar convenience functions, so that database functions may check if messages have been added by other functions. For example `get_notifications()` or `get_notification(type)` or even `has_notification_error()` could be used by a function to check if another called function added a notification of type error that should cause the transaction to abort or short-circuit.

Of course, there is the question of what an "ambient" transaction-specific context is, and how it would be implemented. By "ambient" I mean a context that can be accessed directly and from any function, that is, without having to pass it explicitly from function to function. In this sense, it would be similar to any global variable or settings, like the ones used to store and access the JWT claims. By transaction-specific, I mean that the context would apply to the currently running transaction only, with functions not having to worry about any transactional concerns.

As for how to implement it, there are several possibilities, but I'll leave that to the experts (I'm looking at you Benjie ;) 

Irrespective of how it's implemented, an universal mechanism to register notifications from any function would be invaluable for complex applications that need to perform non-trivial business logic. It would allow functions to keep their signatures intact (no need to pass notifications messages back and forth explicitly) while still having the ability to return any relevant notifications to the GraphQL client.

Here's some potential implementations @benjie / @demianuco discussed:

### Sending messages with LISTEN/NOTIFY

One possibility we've discussed is using PostgreSQL Listen/Notify, in which the add_notification() function would _notify_ in a channel previously created (and subscribed to) by PostGraphile, which would "accumulate" notifications thus and ultimately append them to the GraphQL response at the end of the response life-cycle. There are two problems with this approach, however, one being that the payload is limited (shouldn't be a problem, as notification messages should be small), and the other being that the mechanism will not survive a transaction rollback, thus making it impossible to send errors back to the client (show stopper!).

### Sending messages with RAISE NOTICE

Another possibility is to using the simpler `RAISE NOTICE USING ERRCODE` statement, which raises an indelible notice that is unaffected by rollbacks, and which PostGraphile can used to accumulate notification messages in a more prosaic way. It is a bit of a hack, in the sense that RAISE NOTICE was not meant for this, and the notification might pollute logs (though this could also be an advantage for development and debugging!). Also, it is perhaps fragile as it relies on the right level of verbosity being configured, with any changes to that verbosity breaking the functionality.

### Sending messages via unlogged table

Another possibility is to use an explicit global table (e.g. `tx_notifications`) to which notification messages would be added. It would obviously have to be transactionally scoped, with a column storing the current transaction ID (perhaps `select txid_current()`?), but with the transactional concerns encapsulated away by the convenience functions, so that client functions could add notification messages without having to know about transactions. The table would have no limitations in terms of payload, but I'm concerned about performance and contention issues, especially because this table would be highly volatile and in constant and concurrent use.  Perhaps the use of an unlogged table is appropriate in this scenario.

**Comment by benjie** [comment]
Both NOTIFY and the unlogged table approach have the issue that messages are "rolled back" if a transaction is rolled back:

> NOTIFY interacts with SQL transactions in some important ways. Firstly, if a NOTIFY is executed inside a transaction, the notify events are not delivered until and unless the transaction is committed. This is appropriate, since if the transaction is aborted, all the commands within it have had no effect, including NOTIFY.
> -- https://www.postgresql.org/docs/current/sql-notify.html

For informational messages, this means if an error occurs the user can only see the error and not the details leading up to the error. This may or may not be acceptable, but it's certainly a trade-off.

Regarding the `RAISE NOTICE` (or even `RAISE DEBUG`) option, things are a little more interesting:

> Whether messages of a particular priority are reported to the client, written to the server log, or both is controlled by the log_min_messages and client_min_messages configuration variables. See Chapter 19 for more information.
> -- https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html

So we can log different things to the server logs vs to the client. That means we could use a particularly verbose level like DEBUG, LOG, INFO or NOTICE:

> Allowed levels are DEBUG, LOG, INFO, NOTICE, WARNING, and EXCEPTION, with EXCEPTION being the default. 

with a particular error code like `GOPHK` (Graphile OPeration HooKs, though if you try and read it it sounds a little different... Maybe we should choose a better abbreviation). We can then capture these notices, interpret them in a particular way (e.g. JSON) and add to the messages system.

I think the unlogged table route should be avoided, it doesn't really have any advantages over LISTEN/NOTIFY, but it does have many disadvantages.

---

Irrespective of how we do this, I don't think it should be a goal for the messages raised by these "ambient" functions to be retrievable by other SQL code that can make decisions based on it. We could support basic deduplication, perhaps, but if something requires the current operation to stop and it cannot be done in an explicit "before" check then it should raise an exception.

---

Shall ponder more.

**Comment by demianuco** [comment]
Agreed with all of the above. Ponder away, as I'm sure that what you will come back with will be good. I also have a feeling that it will be used a lot.

**Comment by benjie** [comment]
From chat with @demianuco, this pseudo-code is the kind of shape I'd expect to apply to a single field. Once someone gets this working and padded out and checked for various issues (like synchronisation issues?) we could look at adding it more generically to this plugin

```js
const { makeWrapResolversPlugin, makeExtendSchemaPlugin, makePluginByCombiningPlugins, gql } = require('graphile-utils');

const WrapPlugin = makeWrapResolversPlugin({
  Mutation: {
    myMutation(resolver, parent, args, context, info) {
      const { pgClient } = context;
      const notices = [];
      const push = msg => notices.push(msg);
      pgClient.on('notice', push);
      try {
        const result = await resolver();
        return {
          ...result,
          notices,
        };
      } finally {
        pgClient.removeListener('notice', push);
      }
    }
  }
});

const ExtendPlugin = makeExtendSchemaPlugin({
  typeDefs: gql`
    extend type MyMutationPayload {
      messages: SOMETHING_HERE
    }
  `,
  resolvers: {
    MyMutationPayload: {
      messages(myMutationResult, args, context, info) {
        const { notices } = myMutationResult; // < Extracted from the return result of makeWrapResolvers above
        return someMangledVersionOf(notices);
      }
    }
  }
};

module.exports = makePluginByCombiningPlugins([WrapPlugin, ExtendPlugin]);
```

**Comment by benjie** [comment]
I read a bit of the PostgreSQL docs, and scanned through the node-postgres code, and I believe that this is reliable and not vulnerable to race conditions. I sponsor @brianc the node-postgres maintainer so I've reached out to him to validate my assumptions but I think we can go ahead with the `NOTICE` route :tada:.

**Comment by benjie** [comment]
Brian confirmed my suspicions :+1:

**Comment by benjie** [comment]
First punt at this implemented in #16; would love some eyes on it.
